

## 1. Barcodes demultiplexing and summary statistics

In this table, all previously generated by Trevor RADICL-seq data is aggregated in the first row (regardless of barcode). It is followed by the latest patch from August, segregated by barcode. The majority of reads (85%) from the most recent data don't contain a full barcode leaving only a handful of valid read pairs for downstream analysis per barcode. The exception is **Gametocyte** which yields many final pairs (48% of all final pairs). When the barcode is present the bridge is found in 11% of the reads (improvement over 9% in previous runs) but when the barcode is absent the bridge is found in only 3% of the reads. PCR duplicates have been removed. However, when the final valid read pairs are counted the multi-mappers are counted separately, that is for example, if a RNA-tag corresponding to TARE has its DNA-tag mapped to 7 different locations then I will count 7 RNA-DNA final pairs instead of 1. I will generate a table where the count is always 1 reflective of the fact we have only 1 RNA tag.


| barcode           | % in all reads | having the bridge | % in final pairs | num valid final pairs | cis proximal (<1kb) | cis distal | trans-contacts |
| :---------------: | -------------- | ----------------- | ---------------- |   ---------------------- | ------------------- | ---------- | -------------- | 
| **Aggregated**    | -              |  between 1% to %9 | 100% | 3,213,988 | 0.3% | 7.6% | 92.1% |
| *Newest August*   |  
| **Schizont** | 0.9% | 11.2% | 1.0%   | 12,819 | 1.5% | 8.2% | 90.2 % |
| **Ring** | 1.8% | 6.1%  | 0.08%   |  1,075 | 5.7% | 10.1% | 84.2% | 
| **Trophozoite** | 2.7% | 8.2%  | 10.6%   |  140,420 | 0.5% | 7.8% | 91.6% |
| **Gametocyte** | 9.3% | 11.7% | 48.1%   |  632,586 | 0.4% | 7.6% | 92.0% |
| **no barcode**               | 85.3% | 3.0% | 40.1%   |  527,630 | 0.2% | 8.2% | 91.6% |
| *ALL August*                | 100% | 3.8%  | 100%  |  1,314,530 | 0.4% | 7.7% | 91.8% |
| *Advanced Barcoding*   |  
| **Schizont** | 1.1% | 13.8% | 1.1%   | 14,358 | 1.5% | 8.2% | 90.2 % |
| **Ring** | 2.2% | 7.5%  | 0.08%   |  1,166 | 5.7% | 10.1% | 84.2% | 
| **Trophozoite** | 3.2% | 8.4%  | 11.6%   |  152,677 | 0.5% | 7.8% | 91.6% |
| **Gametocyte** | 13.4% | 13.1% | 60.2%   |  792,245 | 0.4% | 7.6% | 92.0% |
| **no barcode**               | 80.1% | 2.1% | 26.9%   |  354,084 | 0.2% | 8.2% | 91.6% |
| *ALL August*                | 100% | 3.8%  | 100%  |  1,314,530 | 0.4% | 7.7% | 91.8% |


Importantly, when looking at the reads that **do have** the bridge, we can also find the barcode: 
| have bridge           | have barcode |
| :---------------: | -------------- | 
| Yes | 58.1% |
| No | 11.4% |


### 1.1 Demultiplexing

Schematically, this is the full setup of a RADICL-seq read.

1]   P5 adapter: AATGATACGGCGACCACCGAGATCT (25 nt)
2]   i5 index: ACACTCTTTCCCTACACGA (19 nt)
3]   Read 1 primer region: CGCTCTTCCGATCT (14 nt)
4]   RNA sequence of interest: unknown (25-27 nt)
5]   Bridge: CTGCTGCTCCTTCCCTTTCCCCTTTTGGTCCGACGGTCCAAGTCAGCAGT (50 nt)
6]   DNA sequence of interest: unknown (25-27 nt)
7]   Read 2 primer region: AGATCGGAAGAGCG (14 nt)
8]   i7 index: GTTCAGCAGGAATGCCGAG (19 nt)
9]   sample barcode: [rev1]GGGGGG / [rev2]ACGGCG / [rev3]AGCAGC / [rev4]TTCTTC (6 nt)
10] P7 adapter: ATCTCGTATGCCGTCTTCTGCTTG (24 nt)

![TARE_dist](figs/legend)

Here is a plot of 200 randomly drawn reads (containing the bridge) coloring their segments. We observe that the i7_index sequences and barcode are present but the P7 adapter could be missing. This is in line with Karine's suggestion to use only the upstream six nt flanking sequence (GCCGAG) + barcode.

![TARE_dist](figs/barcode_in.png)


Those not having the bridge:
![TARE_dist](figs/barcode_no.png)

## 2. Heatmaps

This is a Hi-C style heatmap of all final valid pairs in a given dataset. The x-axis is the DNA genomic axes, and the y-axis is the RNA one. I generated .mcool files which can be easily visualized by [Higlass](https://higlass.io/). Below is a static snapshot image, the .mcool files are in [files](files/) directory.

There are a couple of prominent horizontal lines that I would like to follow up on ([the lines](#21-The-lines)). I will send their locations to Trevor and also check if they match the ones we saw in iMARGI data. This will be very interesting. 

I have also downsampled the data to get a better visual understanding of how much coverage we need: [downsampling](#5-Downsampling).


#### Aggregated data:

![TARE_dist](figs/agg.png)



#### Newest August:

![TARE_dist](figs/nc.png)


#### Per Barcode (Newest August):

We have very few reads per barcode in some cases.

| **Schizont**   | **Ring**  |
| :--: | :--: |
| ![](figs/n1.png) | ![](figs/n2.png) |
| **Trophozoite**   | **Gametocyte**  |
| ![](figs/n3.png) | ![](figs/n4.png) |


### 2.1. The lines
We have previously observed several prominent RNA lines in the iMARGI data and speculated a bit about their possible biological interpretation. In the RADICL data above, we do see again a couple of such horizontal lines. The exact coordinates of the main ones are:

chr7: 1084000 - 1091000
chr5: 1290000 - 1297000
chr1: 474000 -  479000

and they match the ones from the iMARGI data. That is very interesting.

Here is a plot of where the RNA tags map across the *Plasmodium* genome. The width of each bin is 1000bp. 

![TARE_dist](figs/rna_arr_radicl.png)

Next, I compared this RNA tag distribution to that in the iMARGi data (normalizing the counts in each bin by the total number of final valid RNA tags in each dataset). 

![TARE_dist](figs/rna_arr_combined.png)




## 3. TARE mappings

Here is our favorite positive control. All reads (multi-mappers) corresponding to TARE have been aggregated in the genomic plots below. While there is noise, the peaks with more than 5 reads are at the telomere regions indicating that the experiment has worked! This is true for both the Aggregated data as well as the latest August patch.    

#### Aggregated data:

![TARE_dist](figs/TARE_agg.png)


#### Per Barcode (Newest August):

| **Schizont**   | **Ring**  |
| :--: | :--: |
| ![](figs/TARE_aug_1.png) | ![](figs/TARE_aug_2.png) |
| **Trophozoite**   | **Gametocyte**  |
| ![](figs/TARE_aug_3.png) | ![](figs/TARE_aug_4.png) |


## 4. Other lncRNAs


First, this is a histogram showing the number of final read pairs that I have for each of the known lncRNAs. For a decent number of them I have more than 200 (TARE has 220) read pairs I can use to investigate their biding pattern. 

![lnc_hist](figs/hist.png)
 
For those with the most reads, here are their genomic tracks: 

![lnc_hist](figs/genome_track.png)

Here is the same plot of tags overlapping known lncRNAs for the August data.

![lnc_hist](figs/aug.png)


## 5. Multi-mappers

Most DNA-tags map uniquely, TARE is an exception.

![lnc_hist](figs/hist_mulitpammers.png)


## 5. Downsampling

I downsampeled the Aggregated data by randomly removing **x%** of the final valid RNA tags (and all of their corresponding DNA tags in case of multimapping.) Consequently, the heatmaps become sparser (whither)

| 50% rna (~1.5m pairs )  | 20% rna (~0.6m pairs )  |
| :--: | :--: |
| ![](figs/ds1.png) | ![](figs/ds2.png) |
| 10% rna (~0.3m pairs )  | 5% rna (~0.15m pairs )  |
| ![](figs/ds3.png) | ![](figs/ds4.png) |


